[MIG] product_cost_usd: module migrated to v15.0 t#53862

agarzaarvizu opened the pull request:
## This PR refers to the task[#53862](https://www.vauxoo.com/web#action=1928&cids=1&id=53862&menu_id=1490&model=project.task&view_type=form)

- product_cost_usd module migrated.

Commit:
[ADD] product_cost_usd: handle cost in USD on products

The approach is to have the USD as the base currency,
so that the sale prices in company currency can be
kept up to date through the exchange rate.

Module features:
- New Cost USD field on the Product form.
- Validate Cost USD so that it is not less than the
  list price of the supplier.
- Avoid save a 'Cost in USD' when product does not has
  assigned a supplier with price in USD.
- Allowed pricelist computation based on Cost in USD.
- Added unit tests to validate constrains and pricelists
  computation.
- Compatibility with module sale_margin.

Commit:
[FIX] product_cost_usd: replace pricelist in unit test

The Default Pricelist might be modified in other modules
by this reason is not convinient use it in unit tests.

Commit:
[FIX] product_cost_usd: use float_compare in unit test

agarzaarvizu commented:
@luisg123v Done, please review

cc @desdelinux

luisg123v reviewed (changes requested):
There's only one commit, you should bring the whole history. Follow instructions depicted [here](https://github.com/OCA/maintainer-tools/wiki/Migration-to-version-15.0#technical-method-to-migrate-a-module-from-140-to-150-branch)

agarzaarvizu commented:
@luisg123v Done

Commit:
[IMP] product_cost_usd: black and isort to the module

Commit:
[FIX] product_cost_usd: fix models, views and manifest

- Rename model files and remove parameters in super
- Rename views files and views IDs
- Change category and fix version in manifest
- Remove headers on files

luisg123v commented on product_cost_usd/models/pricelist.py:
```diff
@@ -0,0 +1,65 @@
+        results = super()._compute_price_rule(
+            products_qty_partner, date=date, uom_id=uom_id)
+        usd_currency = self.env.ref('base.USD')
+        date = fields.Date.today()
```
`context_today`?

luisg123v commented on product_cost_usd/models/pricelist.py:
```diff
@@ -0,0 +1,65 @@
+
+    base = fields.Selection(
+        selection_add=[('standard_price_usd', 'Cost in USD')],
+        ondelete={'standard_price_usd': 'set default', 'Cost in USD': 'set default'})
```
No such option "Cost in USD", there's only one

luisg123v commented on product_cost_usd/models/sale_order.py:
```diff
@@ -0,0 +1,49 @@
+        based on cost in usd.
+        """
+        res = super()._compute_purchase_price()
+        date = self.order_id.date_order
```
`for line in self:`

luisg123v commented on product_cost_usd/models/sale_order.py:
```diff
@@ -0,0 +1,49 @@
+        calculate the purchase price when pricelist item is based on cost
+        in usd.
+        """
+        super()._compute_margin()
```
I know returning value from a compute method is not important, but it's harmless and don't doing it triggers a lint

luisg123v commented on product_cost_usd/models/sale_order.py:
```diff
@@ -0,0 +1,49 @@
+        prices = self._compute_purchase_price()
+        if prices:
+            return prices.get('purchase_price')
+        return
```
Superfluous return, you're not returning anything, which is the default value when no return is used

luisg123v commented on product_cost_usd/tests/test_standard_price_usd.py:
```diff
@@ -0,0 +1,99 @@
+
+    def test_03(self):
+        """ Test constraint check_cost_and_price. """
+        self.product.seller_ids = None
```
`False` or `commands.clear()`

luisg123v commented on product_cost_usd/tests/test_standard_price_usd.py:
```diff
@@ -0,0 +1,99 @@
+            float_compare(product.price, expected_price, precision_digits=2),
+            0, "Product price should be %s" % product.price)
+
+    def test_03(self):
```
test method names should be meaningful

luisg123v commented on product_cost_usd/tests/test_standard_price_usd.py:
```diff
@@ -0,0 +1,99 @@
+                'You cannot create or modify a product if the cost in USD'):
+            self.set_standard_price_usd(1)
+
+    def test_sale_margin_normal(self):
```
`test_0*`

luisg123v commented on product_cost_usd/tests/test_standard_price_usd.py:
```diff
@@ -63,51 +63,19 @@ def test_02(self):
                 'You cannot create or modify a product if the cost in USD'):
             self.set_standard_price_usd(1)
 
-    def test_sale_margin(self):
```
Why are you removing this test?

agarzaarvizu commented:
@luisg123v Done

luisg123v commented on product_cost_usd/models/sale_order_line.py:
```diff
@@ -0,0 +1,45 @@
+import pdb
```
Remove

luisg123v commented on product_cost_usd/models/sale_order_line.py:
```diff
@@ -0,0 +1,45 @@
+            purchase_price = line.product_id.standard_price_usd
+            if line.product_uom != line.product_id.uom_id:
+                purchase_price = line.product_id.uom_id._compute_price(purchase_price, line.product_uom)
+            line.purchase_price = frm_cur._convert(purchase_price, to_cur, self.env.company, date, round=False)
```
Company from SO, not from env

luisg123v commented on product_cost_usd/models/sale_order_line.py:
```diff
@@ -0,0 +1,45 @@
+            date = line.order_id.date_order
+            price_rule = line.order_id.pricelist_id._compute_price_rule([(line.product_id, 1, False)])
+            price, rule = price_rule[line.product_id.id]
+            suitable_rule = line.order_id.pricelist_id.item_ids.filtered(lambda x: x.id == rule)
```
Since pricelist and SO is used in several places, it would be convenient to save then in variables

luisg123v commented on product_cost_usd/models/sale_order_line.py:
```diff
@@ -0,0 +1,45 @@
+            if line.product_uom != line.product_id.uom_id:
+                purchase_price = line.product_id.uom_id._compute_price(purchase_price, line.product_uom)
+            line.purchase_price = frm_cur._convert(purchase_price, to_cur, self.env.company, date, round=False)
+            return res
```
This is inside a loop, but you're returning, so only the first record is being considered

luisg123v commented on product_cost_usd/models/sale_order_line.py:
```diff
@@ -0,0 +1,45 @@
+            suitable_rule = line.order_id.pricelist_id.item_ids.filtered(lambda x: x.id == rule)
+            if not suitable_rule or suitable_rule.base != "standard_price_usd":
+                return res
+            frm_cur = self.env.ref("base.USD")
```
`currency_usd`

luisg123v commented on product_cost_usd/tests/test_standard_price_usd.py:
```diff
@@ -0,0 +1,148 @@
+            self.set_standard_price_usd(1)
+
+    def test_05_sale_margin(self):
+        """Test the sale margin module using a pricelist with cost in usd."""
```
```suggestion
        """Test the sale margin module using a pricelist with cost in USD."""
```

luisg123v commented on product_cost_usd/tests/test_standard_price_usd.py:
```diff
@@ -0,0 +1,148 @@
+            {
+                "date_order": fields.Datetime.now(),
+                "name": "Test",
+                "order_line": [
```
Use the `Form` class, please.

Same below

luisg123v commented on product_cost_usd/tests/test_standard_price_usd.py:
```diff
@@ -0,0 +1,148 @@
+        """Test USD pricelist based on cost in usd."""
+        self.set_standard_price_usd(880)
+        product = self.product.with_context(pricelist=self.pricelist_15_usd.id)
+        expected_price = float_round(product.standard_price_usd * 1.15, precision_rounding=self.usd.rounding)
```
Now the currency has a `currency.round(amount)` method

luisg123v commented:
@agarzaarvizu task ID is wrong in both title and description

rolandojduartem commented:
Hi, @luisg123v and @agarzaarvizu, I have bad news :(. I added this module in tanner_common and I had [these](https://gist.github.com/rolandojduartem/3ecb74bd35bead24d3650e6466008df9) tracebacks using it, indeed, you may see that if I do not add this module the tests are working well [here](https://git.vauxoo.com/vauxoo/tanner-common/-/merge_requests/18)

agarzaarvizu commented:
@luisg123v Finally done!! Hahaha please review.

cc @rolandojduartem

Commit:
[FIX] product_cost_usd: inheritance errors related to the old view id

- add pre-migration in order to remove fix errors on inherited views of_product_template_tann_inhrt view

luisg123v commented on product_cost_usd/models/sale_order_line.py:
```diff
@@ -0,0 +1,29 @@
+        """
+        res = super()._compute_purchase_price()
+        for line in self:
+            pricelist_id = line.order_id.pricelist_id
```
```suggestion
            pricelist = line.order_id.pricelist_id
```

luisg123v commented on product_cost_usd/models/sale_order_line.py:
```diff
@@ -0,0 +1,29 @@
+            price, rule = price_rule[line.product_id.id]
+            suitable_rule = pricelist_id.item_ids.filtered(lambda x: x.id == rule)
+            if not suitable_rule or suitable_rule.base != "standard_price_usd":
+                return res
```
I think you meant `continue`

luisg123v commented on product_cost_usd/tests/test_standard_price_usd.py:
```diff
@@ -0,0 +1,117 @@
+        self.product.write({"standard_price_usd": price})
+
+    def test_01_usd_pricelist(self):
+        """Test USD pricelist based on cost in usd."""
```
```suggestion
        """Test USD pricelist based on cost in USD."""
```

luisg123v commented on product_cost_usd/tests/test_standard_price_usd.py:
```diff
@@ -0,0 +1,117 @@
+        self.partner = self.env.ref("base.res_partner_4")
+        self.product_uom = self.env.ref("uom.product_uom_unit")
+        self.product = self.env.ref("product.product_product_24")
+        self.pricelist_15_usd = self.env["product.pricelist"].create(
```
I'd recommend moving this to data demo, otherwise this will be created on every test

luisg123v commented on product_cost_usd/i18n/es.po:
```diff
@@ -0,0 +1,65 @@
+#
+msgid ""
+msgstr ""
+"Project-Id-Version: Odoo Server 10.0+e\n"
```
```suggestion
"Project-Id-Version: Odoo Server 15.0+e\n"
```

I'd recomment re-exporting translations

luisg123v commented on product_cost_usd/models/product_pricelist.py:
```diff
@@ -0,0 +1,47 @@
+
+        Why this inheritance?
+
+        This method always compute the product price, from product currency
```
```suggestion
        This method always computes the product price, from product currency
```

luisg123v commented on product_cost_usd/models/product_pricelist.py:
```diff
@@ -0,0 +1,47 @@
+        This method always compute the product price, from product currency
+        (mostly the same company currency) into pricelist currency. When the
+        pricelist item is based on cost in USD is necessary that currency
+        conversion is made from USD currency. By this reason we must go back
```
```suggestion
        conversion is made from USD currency. For this reason, we must go back
```

luisg123v commented on product_cost_usd/models/product_template.py:
```diff
@@ -0,0 +1,42 @@
+        usd_currency = self.env.ref("base.USD")
+        prec = self.env["decimal.precision"].precision_get("Product Price")
+        usd_seller = self.seller_ids.filtered(lambda x: x.currency_id == usd_currency)
+        list_price = usd_seller.price if usd_seller else 0.0
```
If seller is not found, it will be an empty recordset, and `empty.float_field` isalready 0.0:

```suggestion
        list_price = usd_seller.price
```

luisg123v commented on product_cost_usd/models/product_template.py:
```diff
@@ -0,0 +1,42 @@
+        standard_price_usd = self.standard_price_usd
+        if not usd_seller and float_compare(standard_price_usd, 0, precision_digits=prec) > 0:
+            raise ValidationError(
+                _("You must have at least one supplier with price in USD" " before assign a Cost in USD")
```
```suggestion
                _("You must have at least one supplier with price in USD before assigning a Cost in USD")
```

luisg123v commented on product_cost_usd/models/product_template.py:
```diff
@@ -0,0 +1,42 @@
+                    "- Supplier list price = %s\n"
+                    "- Cost in USD = %s"
+                )
+                % (list_price, standard_price_usd)
```
Use new translation syntax

luisg123v commented on product_cost_usd/models/product_template.py:
```diff
@@ -0,0 +1,42 @@
+        """
+        usd_currency = self.env.ref("base.USD")
+        prec = self.env["decimal.precision"].precision_get("Product Price")
+        usd_seller = self.seller_ids.filtered(lambda x: x.currency_id == usd_currency)
```
Take into account constraints can get a recordset, so I think you should iterate over the result of self to perform validations

luisg123v commented on product_cost_usd/views/product_template_views.xml:
```diff
@@ -0,0 +1,14 @@
+<?xml version="1.0" encoding="utf-8"?>
```
It seems prettier wasn't run for this module

```suggestion
<?xml version="1.0" encoding="utf-8" ?>
```

Same in below cases

luisg123v commented on product_cost_usd/views/product_template_views.xml:
```diff
@@ -0,0 +1,14 @@
+        </field>
+    </record>
+</odoo>
+
```
Remove blank line

luisg123v commented on product_cost_usd/i18n/es.po:
```diff
@@ -0,0 +1,65 @@
+#. module: product_cost_usd
+#: model:ir.model.fields,field_description:product_cost_usd.field_product_template_standard_price_usd
+msgid "Cost in USD"
+msgstr "Costo en Dólares"
```
```suggestion
msgstr "Costo en USD"
```

luisg123v commented on product_cost_usd/i18n/es.po:
```diff
@@ -0,0 +1,65 @@
+"\n"
+"- Supplier list price = %s\n"
+"- Cost in USD = %s"
+msgstr "No puede crear o modificar un producto si el costo en USD es menor que el precio del proveedores..\n"
```
```suggestion
msgstr "No puede crear o modificar un producto si el costo en USD es menor que el precio del proveedor.\n"
```

luisg123v commented on product_cost_usd/i18n/es.po:
```diff
@@ -0,0 +1,65 @@
+#: code:addons/product_cost_usd/models/product.py:33
+#, python-format
+msgid "You must have at least one supplier with price in USD before assign a Cost in USD"
+msgstr "Debe tener al menos un proveedor con precio en Dólares antes de asignar un Costo en Dólares a el producto"
```
```suggestion
msgstr "Debe tener al menos un proveedor con precio en Dólares antes de asignar un Costo en Dólares al producto"
```

luisg123v reviewed (changes requested):
I think some of these can be applied directly in v11

Commit:
[IMP] product_cost_usd: fixing translations and adding demo data

- Fixing translations
- Fixing code comments on models
- Adding demo data for product_pricelist

Commit:
[MIG] product_cost_usd: module migrated to v15.0

agarzaarvizu commented:
@luisg123v Done, please review

luisg123v commented on product_cost_usd/migrations/11.0.1.0.2/pre-migration.py:
```diff
@@ -0,0 +1,13 @@
+def delete_inherited_views(cr):
```
One of the migration tasks is to remove migration scripts

luisg123v commented on product_cost_usd/models/product_template.py:
```diff
@@ -0,0 +1,41 @@
+        """
+        usd_currency = self.env.ref("base.USD")
+        for product in self:
+            prec = product.env["decimal.precision"].precision_get("Product Price")
```
Decimal precision doesn't depennd on the product

luisg123v commented on product_cost_usd/models/product_template.py:
```diff
@@ -0,0 +1,41 @@
+                        " is less than the supplier list price.\n\n"
+                        "- Supplier list price = %s\n"
+                        "- Cost in USD = %s",
+                        list_price, standard_price_usd
```
This semmseems to not be the black format:

```suggestion
                        list_price,
                        standard_price_usd,
```

luisg123v commented on product_cost_usd/i18n/es.po:
```diff
@@ -0,0 +1,96 @@
+#. module: product_cost_usd
+#: model:ir.model,name:product_cost_usd.model_sale_order_line
+msgid "Sales Order Line"
+msgstr "Línea de la orden de venta"
```
pedido

agarzaarvizu commented:
@luisg123v Done

rolandojduartem commented:
Hi, @agarzaarvizu, I just want to say I'm getting this traceback, I'm just running the check_cost_and_price method.

Traceback (most recent call last):
  File "/.repo_requirements/odoo/odoo/models.py", line 5199, in ensure_one
    _id, = self._ids
ValueError: too many values to unpack (expected 1)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/odoo/build/vauxoo/tanner-common/tanner_common/tests/test_product_template.py", line 35, in test_02_check_cost_and_price                                                        
    self.product.product_tmpl_id.check_cost_and_price()
  File "/home/odoo/build/vauxoo/tanner-common/tanner_common/models/product_template.py", line 24, in check_cost_and_price                                                                    
    return super().check_cost_and_price()
  File "/home/odoo/addons-vauxoo/product_cost_usd/models/product_template.py", line 26, in check_cost_and_price                                                                              
    list_price = usd_seller.price
  File "/.repo_requirements/odoo/odoo/fields.py", line 1050, in __get__
    record.ensure_one()
  File "/.repo_requirements/odoo/odoo/models.py", line 5202, in ensure_one
    raise ValueError("Expected singleton: %s" % self)
ValueError: Expected singleton: product.supplierinfo(3, 1, 2)

cc @luisg123v

luisg123v commented:
I've rebased this branch so tests may run

luisg123v commented on product_cost_usd/models/product_template.py:
```diff
@@ -0,0 +1,42 @@
+        usd_currency = self.env.ref("base.USD")
+        prec = self.env["decimal.precision"].precision_get("Product Price")
+        for product in self:
+            usd_seller = product.seller_ids.filtered(lambda x: x.currency_id == usd_currency)
```
```suggestion
            usd_seller = product.seller_ids.filtered(lambda x: x.currency_id == usd_currency)[:1]
```

But maybe `_select_seller` may be used.

agarzaarvizu commented:
@luisg123v Done

desdelinux commented:
@agarzaarvizu @luisg123v No aparece el tag del campo en la vista del Producto, sólo aparece el valor:

![image](https://user-images.githubusercontent.com/50924836/188227037-3dbf4e0f-cdfc-4b1d-bb02-57f41c3e234d.png)

agarzaarvizu commented:
@luisg123v @desdelinux Done

desdelinux commented:
LGTM
